cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller_test.go

```diff
@@ -371,3 +394,533 @@ func TestMachinesInMachineSet(t *testing.T) {
+			Name:      "machinedeployment",
+			Namespace: "test-namespace",
+		},
+	}
+
+	for i, tc := range []struct {
+		description string
+		annotations map[string]string
+		errors      bool
+		nodegroups  int
+	}{{
+		description: "errors with bad annotations",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "-1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		nodegroups: 0,
+		errors:     true,
+	}, {
+		description: "success but nodegroup count is 0 as deployment has no bounds",
+		nodegroups:  0,
+		errors:      false,
+	}, {
+		description: "success with valid bounds",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		nodegroups: 1,
+		errors:     false,
+	}} {
+		t.Logf("test #%d: %s", i, tc.description)
+
+		machineDeployment := machineDeploymentTemplate.DeepCopy()
+		machineDeployment.Annotations = tc.annotations
+
+		controller, stop := mustCreateTestController(t, testControllerConfig{
+			machineObjects: []runtime.Object{
+				machineDeployment,
+			},
+		})
+		defer stop()
+
+		nodegroups, err := controller.machineDeploymentNodeGroups()
+		if tc.errors && err == nil {
+			t.Errorf("expected an error")
+		}
+
+		if !tc.errors && err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+
+		if tc.errors && nodegroups != nil {
+			t.Fatalf("test case logic error")
+		}
+
+		if actual := len(nodegroups); actual != tc.nodegroups {
+			t.Errorf("expected %d, got %d", tc.nodegroups, actual)
+		}
+	}
+}
+
+func TestControllerNodeGroupsWithMachineSets(t *testing.T) {
+	machineSetOwnedByMachineDeployment := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset-owned-by-deployment",
+			Namespace: "test-namespace",
+			OwnerReferences: []v1.OwnerReference{{
+				Kind: "MachineDeployment",
+				Name: "machinedeployment",
+			}},
+		},
+	}
+
+	machineDeployment := &v1beta1.MachineDeployment{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineDeployment",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machinedeployment",
+			Namespace: "test-namespace",
+			Annotations: map[string]string{
+				nodeGroupMinSizeAnnotationKey: "1",
+				nodeGroupMaxSizeAnnotationKey: "10",
+			},
+		},
+	}
+
+	machineSetTemplate := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset",
+			Namespace: "test-namespace",
+		},
+	}
+
+	for i, tc := range []struct {
+		description string
+		annotations map[string]string
+		errors      bool
+		nodegroups  int
+	}{{
+		description: "errors with bad annotations",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "-1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		nodegroups: 0,
+		errors:     true,
+	}, {
+		description: "success but nodegroup count is 0 as machineset no bounds",
+		nodegroups:  0,
+		errors:      false,
```

Comment:
nit: no need to set it, as long as it's explicit from the context missing errors means no error.

Reply:
That's what I was largely trying to avoid; by being explicit you don't have to guess my (or the tests) intentions.